hw-model/src/lib.rs

```diff
@@ -330,44 +341,107 @@ pub trait HwModel {
+            remaining = &remaining[4..];
+        }
+        if !remaining.is_empty() {
+            let mut word_bytes = [0u8; 4];
+            word_bytes[..remaining.len()].copy_from_slice(remaining);
+            self.soc_mbox()
+                .datain()
+                .write(|_| u32::from_le_bytes(word_bytes));
+        }
 
-        let word_size = RvSize::Word as usize;
-        let remainder = firmware.len() % word_size;
-        let n = firmware.len() - remainder;
+        // Ask the microcontroller to execute this command
+        self.soc_mbox().execute().write(|w| w.execute(true));
 
-        for idx in (0..n).step_by(word_size) {
-            let val = u32::from_le_bytes(firmware[idx..idx + word_size].try_into().unwrap());
-            self.soc_mbox().datain().write(|_| val);
+        // Wait for the microcontroller to finish executing
+        while self.soc_mbox().status().read().status().cmd_busy() {
+            self.step();
+        }
+        let status = self.soc_mbox().status().read().status();
+        if status.cmd_failure() {
+            writeln!(self.output().logger(), ">>> mbox cmd response: failed").unwrap();
+            self.soc_mbox().execute().write(|w| w.execute(false));
+            return Err(ModelError::MailboxCmdFailed);
+        }
+        if status.cmd_complete() {
+            writeln!(self.output().logger(), ">>> mbox cmd response: success").unwrap();
+            self.soc_mbox().execute().write(|w| w.execute(false));
+            return Ok(None);
+        }
+        if !status.data_ready() {
+            return Err(ModelError::UnknownCommandStatus(status as u32));
         }
 
-        // Handle the remainder bytes.
-        if remainder > 0 {
-            let mut last_word = firmware[n] as u32;
-            for idx in 1..remainder {
-                last_word |= (firmware[n + idx] as u32) << (idx << 3);
-            }
-            self.soc_mbox().datain().write(|_| last_word);
+        let mut result = vec![];
+        let mut dlen = self.soc_mbox().dlen().read();
+
+        writeln!(
+            self.output().logger(),
+            ">>> mbox cmd response data ({dlen} bytes)"
+        )
+        .unwrap();
+        while dlen >= 4 {
+            result.extend_from_slice(&self.soc_mbox().dataout().read().to_le_bytes());
+            dlen -= 4;
+        }
+        if dlen > 0 {
+            // Unwrap cannot panic because dlen is less than 4
+            result.extend_from_slice(
+                &self.soc_mbox().dataout().read().to_le_bytes()[..usize::try_from(dlen).unwrap()],
+            );
         }
 
-        // Set Execute Bit
-        self.soc_mbox().execute().write(|w| w.execute(true));
+        self.soc_mbox().execute().write(|w| w.execute(false));
+        // mbox_fsm_ps isn't updated immediately after execute is cleared (!?),
+        // so step an extra clock cycle to wait for fm_ps to update
+        self.step();
```

Comment:
I assume you're seeing this in verilator, not the sw emulator?

Reply:
Yes